src/transformers/models/kosmos2/modeling_kosmos2.py

```diff
@@ -0,0 +1,2106 @@
+            shift_labels = labels[..., 1:].contiguous()
+            batch_size, seq_length, vocab_size = shift_logits.shape
+            # Flatten the tokens
+            loss_fct = CrossEntropyLoss()
+            loss = loss_fct(
+                shift_logits.view(batch_size * seq_length, vocab_size), shift_labels.view(batch_size * seq_length)
+            )
+
+        if not return_dict:
+            output = (lm_logits,) + outputs[1:]
+            return (loss,) + output if loss is not None else output
+
+        return CausalLMOutputWithCrossAttentions(
+            loss=loss,
+            logits=lm_logits,
+            past_key_values=outputs.past_key_values,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+            cross_attentions=outputs.cross_attentions,
+        )
+
+    def prepare_inputs_for_generation(
+        self,
+        input_ids,
+        image_features=None,
+        image_features_mask=None,
+        past_key_values=None,
+        attention_mask=None,
+        use_cache=None,
+        **model_kwargs,
+    ):
+        input_shape = input_ids.shape
+        # if model is used as a decoder in encoder-decoder model, the decoder attention mask is created on the fly
+        if attention_mask is None:
+            attention_mask = input_ids.new_ones(input_shape)
+
+        position_ids = None
+
+        # cut input_ids if past_key_values is used
+        if past_key_values is not None:
+            position_ids = create_position_ids_from_input_ids(
+                input_ids,
+                padding_idx=self.config.pad_token_id,
+                past_key_values_length=0,
+            )[:, -1:]
+
+            input_ids = input_ids[:, -1:]
+            # the image info. is already encoded into the past keys/values
+            image_features = None
+            image_features_mask = None
+        elif image_features_mask is not None:
+            # appending `False` to `image_features_mask` (because `input_ids` grows during generation)
+            batch_size, seq_len = input_ids.size()
+            mask_len = image_features_mask.size()[-1]
+            image_features_mask = torch.cat(
+                (
+                    image_features_mask,
+                    torch.zeros(size=(batch_size, seq_len - mask_len), dtype=torch.bool, device=input_ids.device),
+                ),
+                dim=1,
+            )
+
+        return {
+            "input_ids": input_ids,
+            "image_features": image_features,
+            "image_features_mask": image_features_mask,
+            "past_key_values": past_key_values,
+            "attention_mask": attention_mask,
+            "position_ids": position_ids,
+            "use_cache": use_cache,
+        }
+
+
+class Kosmos2ImageToTextConnector(nn.Module):
+    """The layer that transforms the image model's output to part of the text model's input (namely, image features)"""
+
+    def __init__(self, config: Kosmos2Config):
```

Comment:
The naming probably comes from the paper but seems a bit odd, let's try to use ...ImageProjection and `image_projection` maybe? (Similarly to what align does!)

Reply:
With the current implementation, I make sure loading a `Kosmos2ForConditionalGeneration` checkpoint to `Kosmos2Model` works. But it won't work if we load `Kosmos2ForConditionalGeneration` checkpoint to text or vision component.

I indeed take the idea from CLIP. However, CLIP is slightly simpler than this text-vision conditional generation model. There is a way to make loading to component work, but it will require another arrangement of internal layer structure.